homeassistant/components/simplisafe/button.py

```diff
@@ -0,0 +1,93 @@
+"""Buttons for the SimpliSafe integration."""
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+
+from simplipy.errors import SimplipyError
+from simplipy.system import System
+
+from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import SimpliSafe, SimpliSafeEntity
+from .const import DOMAIN
+from .typing import SystemType
+
+
+@dataclass
+class SimpliSafeButtonDescriptionMixin:
+    """Define an entity description mixin for SimpliSafe buttons."""
+
+    push_action: Callable[[System], Awaitable]
```

Comment:
Newbie question: Do we really need to define a custom Mixin for this?  There isn't a command button mixin that we can use instead?

Reply:
There are a few examples like the `bmw_connected_drive/button.py` that don't bother with the mixin for such a simple scenario and instead just add the property to the derived class:

```python
# Snipped some extra fields
@dataclass
class BMWButtonEntityDescription(ButtonEntityDescription):
    """Class describing BMW button entities."""

    remote_function: Callable[
        [MyBMWVehicle], Coroutine[Any, Any, RemoteServiceStatus]
    ] | None = None
```

It feels like just doing this would be cleaner, and if there's could be a pre-defined type for `Callable[[System], Awaitable]` it feels like that could make it even cleaner:

```python
@dataclass
class SimpliSafeButtonDescription(ButtonEntityDescription):
    """Describe a SimpliSafe button description."""

   push_action: Callable[[System], Awaitable] # <-- Or push_action: ButtonPushAction if it was added?
```

Reply:
The BMW example works because the added attribute is optional – since `ButtonEntityDescription` already defines optional attributes, it's okay for another optional attribute to follow.

In this case, the added attribute is required – if I attempted to extend `ButtonEntityDescription`, I would get a `TypeError`. Per the [spec](https://peps.python.org/pep-0557/#id7):

> TypeError will be raised if a field without a default value follows a field with a default value. This is true either when this occurs in a single class, or as a result of class inheritance.

Reply:
Ah, thanks for the info!